### PR TITLE
Add verification of files for git-extract

### PR DESCRIPTION
**Pull request by kevgo**
@charlierudolph @allewun @ricmatsui 

Implements parts of #420 

* explicitely specifies which files are in which branch after the operation is done
* except for cases where this is very obvious from the commits. In this case the file verification is inlined with the commit verification
* if no file content is given, uses a default file content (file name + " content")
* dries up existing file content verification tables thanks to using Mortadella now
* sorts file tables by 
  * branch name: main first, then alphabetically
  * file name

What do you think?

### CODE REVIEW

File: features/support/files_helpers.rb [hunk 1]
```diff
@@ -13,13 +15,13 @@ def files_in branch:
 # Returns a table of all files in all branches.
 # This is for comparing against expected files in a Cucumber table.
 def files_in_branches
-  result = [%w(BRANCH NAME CONTENT)]
+  result = Mortadella.new headers: %w(BRANCH NAME CONTENT), dry: ['BRANCH']
   existing_local_branches(order: :main_first).each do |branch|
-    files_in(branch: branch).each do |file|
+    files_in(branch: branch).sort.each do |file|
```
Comment (charlierudolph) [inline features/support/files_helpers.rb]: files are automatically sorted, sort should be unnecessary

File: features/support/commit_helpers.rb [hunk 2]
```diff
@@ -67,20 +67,33 @@ def create_commits commits_array
...
-  }
+    branch: current_branch_name,
+    file_content: commit_data[:file_content]
+  )
+  if result[:file_content].blank?
+    result[:file_content] = default_file_content_for commit_data[:file_name]
+  end
+  result
+end
+# rubocop:enable MethodLength
```
Comment (charlierudolph) [inline features/support/commit_helpers.rb]: Please stop adding these.
If needed lets bump to allowed method length, but otherwise please find a way to meet the standard.
Here it can be done with 
```ruby
def add_default_commit_data commit_data
  file_name = "default file name #{SecureRandom.urlsafe_base64}"
  file_content = default_file_content_for(commit_data[:file_name] or file_name)
  commit_data.clone.reverse_merge(
    file_name: file_name,
    message: 'default commit message',
    location: 'local and remote',
    branch: current_branch_name,
    file_content: file_content
  )
end
```

File: features/support/files_helpers.rb (hunk 1 above)
Comment (kevgo) [inline features/support/files_helpers.rb]: I wondered about that, but you seem to be right. Removed.

File: features/support/commit_helpers.rb (hunk 2 above)
Comment (kevgo) [inline features/support/commit_helpers.rb]: Yeah, those Rubocop-exceptions really aren't good. Thanks for the improvement! It works (although it shouldn't from looking at it), so I'm adopting your solution for now. I have improved it even further to 

```ruby
def add_default_commit_data commit_data
  file_name = "default file name #{SecureRandom.urlsafe_base64}"
  commit_data.clone.reverse_merge(
    file_name: file_name,
    message: 'default commit message',
    location: 'local and remote',
    branch: current_branch_name,
    file_content: default_file_content_for(commit_data[:file_name] || file_name)
  )
end
```

**Comment by kevgo** [comment]
Sorry for the intermittent brokenness here, it took me by surprise. In the future I'll check on CircleCI before even opening a PR. Everything works now, and all comments addressed.

File: features/git-extract/cherry_pick_conflict/with_remote_origin/with_open_changes.feature [hunk 3]
```diff
@@ -85,6 +80,13 @@ Feature: git extract: resolving conflicts between main branch and extracted comm
       |          |                  | refactor commit | conflicting_file |
       | refactor | local and remote | main commit     | conflicting_file |
       |          |                  | refactor commit | conflicting_file |
+    And now I have the following committed files
+      | BRANCH   | NAME             | CONTENT          |
+      | main     | conflicting_file | main content     |
+      | feature  | conflicting_file | refactor content |
+      | feature  | feature_file     |                  |
+      | refactor | conflicting_file | resolved content |
```
Comment (charlierudolph) [inline features/git-extract/cherry_pick_conflict/with_remote_origin/with_open_changes.feature]: When we have this, I don't think its necessary to have the files in the commit tables

Comment (kevgo) [inline features/git-extract/cherry_pick_conflict/with_remote_origin/with_open_changes.feature]: Interesting idea. In this case we should create separate steps for verifying commits and verifying commits and files

```cucumber
    Then now I have the following commits
      | BRANCH   | LOCATION         | MESSAGE     | 
      | main     | local and remote | main commit | 
      (does not verify the files in the commits)

    Then now I have the following commits and files
      | BRANCH  | LOCATION         | MESSAGE        | FILE NAME        | FILE CONTENT     |
      | main    | local and remote | main commit    | conflicting_file | resolved content |
      | feature | local and remote | feature commit |                  |                  |
      (verifies the file content, backfilling with default values if necessary)

Comment (kevgo) [inline features/git-extract/cherry_pick_conflict/with_remote_origin/with_open_changes.feature]: I implemented your idea, and removed all file data from the commit tables. Looks good!

**Comment by kevgo** [comment]
all comments addressed

File: features/git-extract/cherry_pick_conflict/with_remote_origin/with_open_changes.feature [hunk 4]
```diff
@@ -79,12 +74,19 @@ Feature: git extract: resolving conflicts between main branch and extracted comm
...
+      | main     | local and remote | main commit     |
+      | feature  | local            | feature commit  |
+      |          |                  | refactor commit |
+      | refactor | local and remote | main commit     |
+      |          |                  | refactor commit |
+    And now I have the following committed files
+      | BRANCH   | NAME             | CONTENT          |
+      | main     | conflicting_file | main content     |
+      | feature  | conflicting_file | refactor content |
+      | feature  | feature_file     |                  |
```
Comment (charlierudolph) [inline features/git-extract/cherry_pick_conflict/with_remote_origin/with_open_changes.feature]: Drying up should be done in a separate PR now that you are using Mortadella for these tables

**Comment by charlierudolph** [comment]
Sorry I'm only coming to this realization now but why can't we just have the file content in the commit table? That way we can keep using the diffs and don't need an extra table

File: features/git-extract/cherry_pick_conflict/with_remote_origin/with_open_changes.feature (hunk 4 above)
Comment (kevgo) [inline features/git-extract/cherry_pick_conflict/with_remote_origin/with_open_changes.feature]: Agreed. This is not a dry up, but using the default value for the file content. Its been there all the time, and is defined this way in the `Given` step.

**Comment by kevgo** [comment]
I didn't understand your last comment. The commit table to define commits already has file contents. The commit table to verify commits can contain file contents, and we use that in simple cases. The problem is that in more complex cases (with merge commits) its not obvious which branch contains which file in the end, and that's something I think should be documented in the feature specs.

**Comment by ricmatsui** [comment]
Looks good to me. It'll definitely look better once the file sections are dried up, too.

**Comment by kevgo** [comment]
Ping

File: features/git-extract/cherry_pick_conflict/with_remote_origin/with_open_changes.feature [hunk 5]
```diff
@@ -79,12 +74,19 @@ Feature: git extract: resolving conflicts between main branch and extracted comm
...
+      | BRANCH   | LOCATION         | MESSAGE         |
+      | main     | local and remote | main commit     |
+      | feature  | local            | feature commit  |
+      |          |                  | refactor commit |
+      | refactor | local and remote | main commit     |
+      |          |                  | refactor commit |
+    And now I have the following committed files
+      | BRANCH   | NAME             | CONTENT          |
+      | main     | conflicting_file | main content     |
+      | feature  | conflicting_file | refactor content |
```
Comment (charlierudolph) [inline features/git-extract/cherry_pick_conflict/with_remote_origin/with_open_changes.feature]: This line can be removed. Below you specify the the conflicting file on the feature branch actually has the content "resolved content"

Comment (charlierudolph) [inline features/git-extract/cherry_pick_conflict/with_remote_origin/with_open_changes.feature]: nevermind, i got confused

**Comment by kevgo** [comment]
Okay, tests fixed. Ready to ship?

File: features/git-extract/multiple_commits/with_remote_origin/with_open_changes.feature [hunk 6]
```diff
@@ -32,11 +32,20 @@ Feature: git extract: extracting multiple commits (with open changes)
...
+      |          |                  | refactor2 commit   |
+    And now I have the following committed files
+      | BRANCH   | NAME             |
+      | main     | remote_main_file |
+      | feature  | feature_file     |
+      | feature  | refactor1_file   |
+      | feature  | refactor2_file   |
+      | refactor | refactor1_file   |
+      | refactor | refactor2_file   |
+      | refactor | remote_main_file |
```
Comment (charlierudolph) [inline features/git-extract/multiple_commits/with_remote_origin/with_open_changes.feature]: I don't see any reason why this file table was broken out. With no content to check against I don't think its necessary

Comment (charlierudolph) [inline features/git-extract/multiple_commits/with_remote_origin/with_open_changes.feature]: To me a file table should only be needed when there are conflicts.

Comment (charlierudolph) [inline features/git-extract/multiple_commits/with_remote_origin/with_open_changes.feature]: Thus same comment applies for the next 7 files

Comment (kevgo) [inline features/git-extract/multiple_commits/with_remote_origin/with_open_changes.feature]: I want to explicitly and ergonomically specify/document which files exist on which branch after the command is run. 

This is relevant information. As a user, I care that my files are in the right places after I run a Git Town command. More so than which commits exist. 

This could in simple cases like this one theoretically be deducted from the commits, but I think that's cumbersome and brittle, and a dedicated list is better. 

Also, there is value in having a uniform way of specifying the files. In a separate thread above we decided to separate commit verification from file verification. I think it makes sense to do this here as well. 

If the file content is generic, then it can (and should) be omitted, to keep the specs simple.

What do you think?

**Comment by kevgo** [comment]
All comments addressed

File: features/git-extract/multiple_commits/with_remote_origin/with_open_changes.feature (hunk 6 above)
Comment (charlierudolph) [inline features/git-extract/multiple_commits/with_remote_origin/with_open_changes.feature]: Okay. That makes sense. 

In thinking about this, I don't think there is any reason to specify files when there are no conflicts (this of course can be dealt with separately)

Comment (kevgo) [inline features/git-extract/multiple_commits/with_remote_origin/with_open_changes.feature]: I think it is in general insightful to see the new file setup after a Git Town command has moved some commits and files around. This spec is a good example. It extracts some commits into another branch. It is useful to see which files the extracted branch contains afterwards, and that these files are also still in the original branch.

If there is a situation where the file setup after a Git Town command is irrelevant, then I'm happy to omit this table. I'd say the whole Git Town command would be relatively irrelevant in this case though, since it didn't do much. :)

Comment (charlierudolph) [inline features/git-extract/multiple_commits/with_remote_origin/with_open_changes.feature]: But if each commit is dealing with a different file then I don't see any reason to inspect the files. I'll spike out an example and we can discuss there.

**Comment by charlierudolph** [comment]
:ship:

File: features/git-extract/multiple_commits/with_remote_origin/with_open_changes.feature (hunk 6 above)
Comment (kevgo) [inline features/git-extract/multiple_commits/with_remote_origin/with_open_changes.feature]: Okay, sounds good.
